Add docs about CAPI node pools

fiunchinho opened the pull request:
### What does this PR do?

Towarsd https://github.com/giantswarm/roadmap/issues/2984
Docs were missing for CAPI node pools. I also renamed current docs to `node-pools-vintage`.

### Have you maintained the front matter?

(Please bump the last_review_date in case this qualifies as a review for an entire page. Provide user_questions which should be answered in the page. Provide a meaningful description.)

calvix commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,304 @@
+result in node pools providing less than the desired number of nodes. The more availability zones a node pool spans,
+the less likely this problem is to occur.
+
+- By default, clusters can have up to 5 node pools. This is limited by the AWS service quota named "IPv4 CIDR blocks
```
this is only true for vintage

calvix replied:
In CAPI we do not add CIDR block for new node pool so the limit do not apply

calvix commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+per VPC" in the [VPC section](https://docs.aws.amazon.com/vpc/latest/userguide/amazon-vpc-limits.html). As the AWS
+account owner, you can request an increase of this limit via the AWS console.
+
+- Node pools can span a maximum of four availability zones. This limit affects both the number of availability zones
```
TBH I am not sure this is true as well, I think we are limited by number of AZs in the region and also there is setting how many AZs should be used for subnets  https://github.com/giantswarm/cluster-aws/blob/master/helm/cluster-aws/templates/_aws_cluster.tpl#L56

AndiDog replied:
Probably this limitation, if it's even applicable, is too detailed and most customers won't hit it. Maybe we want to remove this hint altogether, or keep only the vintage specifics on a different tab.

calvix commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+nodes, it cannot be unassigned from that cluster. It will remain assigned even if there are no more node pools using
+that availability zone.
+
+    - **Example:** The control plane node(s) are in availability zone A. Node pool 1 uses availability zones B and C. Node pool 2 uses
```
This is only true for vintage

calvix commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+## Definition
+
+A node pool is a set of nodes within a Kubernetes cluster that share the same configuration (machine type, CIDR range, etc.). Each node in the pool is labeled by the node pool's name.
```
` CIDR range` I would remove this as all nodes in whole cluster share same CIDR range,

Different CIDR range per nodepool was true for vintage

calvix commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+- Availability zone distribution
+- Scaling configuration (number of nodes)
+
+A node pool is identified by a name that you can pick as a cluster administrator. The name of the node pool is set as a label on all nodes belonging to the pool.
```
This is a good place to mention limitation of the nodepool name, like how many chars and waht chars are allowed + few examples

AndiDog replied:
And let's also be specific. The introduction already talked about name labeling. Let's reference the actual `giantswarm.io/machine-pool=...` label here. Or since you already specify it below in the per-provider section, maybe rather remove this sentence here (see also suggestion above to add it as bullet point).

calvix commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+## Configuration
+
+Node pools can be created, deleted or updated by changing the configuration used when creating the cluster using [`kubectl`]({{< relref "/getting-started/create-workload-cluster" >}})
```
`kubectl` -> `kubectl gs`

AndiDog commented:
Comment from standup: let's note down that renaming node pools is discouraged because Helm deletes the old one and adds the new one (different name = different object; not a real rename). We should recommend adding a new pool, waiting for it to be healthy, and then removing the old one.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+linkTitle: Node pools (CAPI)
+title: Node pools (CAPI)
+description: A general description of node pools as a concept, its benefits, and some details you should be aware of.
+weight: 50
```
```suggestion
weight: 45
```

to be before vintage for sure (not sure what Hugo does for equal weight).

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+- Machine type
+- Availability zone distribution
+- Scaling configuration (number of nodes)
```
```suggestion
- Scaling configuration (number of nodes)
- Node labels (the pool name is added as node label by default)
```

and then below, we explain that these can be used to run workloads on specific node pools (high RAM, GPU support, specific AZ, etc.)

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+  --machine-pool-name pool0 \
+  --machine-pool-min-size 3 \
+  --machine-pool-max-size 5 \
+  --machine-pool-instance-type r6i.xlarge
```
This looks like an AWS instance type, not Azure VM type.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
````diff
@@ -0,0 +1,322 @@
+```nohighlight
+kubectl get nodes \
+  -o=jsonpath='{range .items[*]}{.metadata.labels.kubernetes\.io/role}{"\t"}{.metadata.labels.giantswarm\.io/machine-deployment}{"\t"}{.metadata.name}{"\n"}{end}' | sort
+master         ip-10-1-5-55.eu-central-1.compute.internal
+worker  a1b2c3-pool0  ip-10-1-6-225.eu-central-1.compute.internal
+worker  a1b2c3-pool0  ip-10-1-6-67.eu-central-1.compute.internal
````
Same as above: these are AWS instance name examples, not Azure VM names

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+The node pool name will be prepended by the cluster name. In the example above, the node pool name will be `a1b2c3-pool0`.
+
+All nodes in the node pool will be labeled with the node pool name, using the  `giantswarm.io/machine-deployment` label.
```
```suggestion
All nodes in the node pool will be labeled with the node pool name, using the `giantswarm.io/machine-deployment` label.
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+The node pool name will be prepended by the cluster name. In the example above, the node pool name will be `a1b2c3-pool0`.
+
+All nodes in the node pool will be labeled with the node pool name, using the  `giantswarm.io/machine-pool` label.
```
```suggestion
All nodes in the node pool will be labeled with the node pool name, using the `giantswarm.io/machine-pool` label.
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+## Assigning workloads to node pools {#assigning-workloads}
+
+Knowing the node pool name of the pool to use, you can use the `nodeSelector` method of assigning pods to the node pool.
```
```suggestion
Knowing the node pool name of the pool to use, you can use the [`nodeSelector` method of assigning pods](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) to the node pool.
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+Knowing the node pool name of the pool to use, you can use the `nodeSelector` method of assigning pods to the node pool.
+
+Assuming that the node pool name is `pool0`, and the cluster name is `a1b2c3`, your `nodeSelector` could for example look like this:
```
I suggest `s/a1b2c3/mycluster/` everywhere, since that name is used in https://docs.giantswarm.io/getting-started/create-workload-cluster/#step-2-template-and-create-the-workload-cluster. Some consistency would be great, and those random names are harder to read.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+{{< /tab >}}
+{{< /tabs >}}
+
+You can assign workloads to node pools in a more indirect way too. This is achieved by using other node attributes which are
+specified via the node pool and which are exposed as node labels.
```
Can we rephrase this so it refers to automatically generated labels? It currently reads as if one needs to explicitly specify them. The given examples such as `topology.kubernetes.io/zone` are auto-generated. We don't need to explain in detail which controller does that, but maybe we can simply refer to well-known labels as listed at https://kubernetes.io/docs/reference/labels-annotations-taints/. Together with the below examples, that's enough.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+You can assign workloads to node pools in a more indirect way too. This is achieved by using other node attributes which are
+specified via the node pool and which are exposed as node labels.
+
+For example: In a case where you have node pools with one instance type. Using a `nodeSelector` with the label `node.kubernetes.io/instance-type` you can assign workloads to matching nodes only.
+
+Another example: In a case where you have different node pools using different availability zones. With a `nodeSelector` using the label `topology.kubernetes.io/zone` you can assign your workload to the nodes in a particular availability zone.
```
```suggestion
For example: you have several node pools with different instance types. Using a `nodeSelector` with the label `node.kubernetes.io/instance-type`, you can assign workloads only to matching nodes.

Another example: you have different node pools using different availability zones. With a `nodeSelector` using the label `topology.kubernetes.io/zone`, you can assign your workload to the nodes in a particular availability zone.
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+  nodepool0:
+    maxSize: 4
+    minSize: 3
+    instanceTypeOverrides:
+    - r6i.xlarge
+    - r5.xlarge
+    - m5.xlarge
```
`nodepool0` should remain the same in the before/after YAML

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
````diff
@@ -0,0 +1,322 @@
+    minSize: 1
+```
+
+All the nodes from each node pool will be labeled accordingly.
+
+## Updating an existing node pool
````
```suggestion
## Updating an existing node pool
```

Not necessary to repeat this, as it was clearly explained above.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+The `kubelet` will wait up to 5 minutes for all pods to terminate and after that it will terminate itself and the shutdown of the AWS EC2 instance will finally proceed.
+
+Please, be aware that it may happen that AWS decides to force terminate the instance before the 5 minutes.
```
```suggestion
Please, be aware that it may happen that AWS decides to force-terminate the instance before the 5 minutes. We recommend draining the respective nodes and moving workloads to other nodes before such an update operation.
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+## Node pool deletion
+
+TBD
```
Until this is solved, maybe we can add a tab for vintage already if we support deletion there (I guess since it's a separate CR without Helm "keep" annotation).

fiunchinho replied:
But we have a different page for `vintage`. The idea is to document vintage on its own page, and use this one only for CAPI.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+TBD
+
+## Using mixed instance types {#mixed-instance-types}
```
```suggestion
## Using mixed instance types (only {{% impl_title "capa_ec2" %}}) {#mixed-instance-types}
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+**Note:** This feature was called ["Similar instance types" on vintage]({{< relref "/advanced/cluster-management/node-pools-vintage#similar-instance-types" >}}).
+
+On AWS EC2 (CAPA) you can override the instance type of the node pool to enable [mixed instances policy on AWS](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_LaunchTemplateOverrides.html).
```
```suggestion
On {{% impl_title "capa_ec2" %}}, you can override the instance type of the node pool to enable [mixed instances policy on AWS](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_LaunchTemplateOverrides.html).
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+## Node pools and autoscaling {#autoscaling}
+
+With node pools, you set the autoscaling range per node pool (suppported on AWS EC2 (CAPA) clusters only). The Kubernetes cluster autoscaler has to decide which node pool to scale under which circumstances.
```
```suggestion
With node pools, you set the autoscaling range per node pool (suppported on {{% impl_title "capa_ec2" %}} clusters only). The Kubernetes cluster autoscaler has to decide which node pool to scale under which circumstances.
```

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+{{< tabs >}}
+{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}
+
+Node pools can make use of [Amazon EC2 Spot Instances](https://aws.amazon.com/ec2/spot/). You just need to enable it and select the max price to pay on the node pool definition.
```
```suggestion
Node pools can make use of [Amazon EC2 Spot Instances](https://aws.amazon.com/ec2/spot/). On the node pool definition, you can enable it and select the maximum price to pay.
```

Avoiding the word "just" because it can also relay a passive-aggressive "Hey it's so easy, a baby could do it" meaning.

AndiDog commented on src/content/advanced/cluster-management/node-pools-capi/index.md:
```diff
@@ -0,0 +1,322 @@
+
+## Limitations
+
+- Clusters without worker nodes (= without node pools) cannot be considered fully functional. In order to have all
+required components scheduled, worker nodes are required. For that reason, we deactivate any monitoring and alerts for
+these clusters and don't provide any proactive support.
+
+### Things to keep in mind
+
+{{< tabs >}}
+{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}
```
```suggestion
{{< tabs >}}
{{< tab id="general" title="General" >}}

- Clusters without worker nodes (= without node pools) cannot be considered fully functional. In order to have all
required components scheduled, worker nodes are required. For that reason, we deactivate any monitoring and alerts for
clusters without worker nodes and don't provide any proactive support.

{{< /tab >}}
{{< tab id="cluster-capa-ec2" for-impl="capa_ec2" >}}
```

AndiDog replied:
and please ensure that all `tab id="..."` are unique (= don't repeat `id="cluster-capa-ec2"` but use `limitations-capa-ec2`) so that the tab-switching logic works i.e. switches all equally-titled tabs at once on the page.

AndiDog commented:
LGTM, but the build still fails (wrong `impl-title`, Markdown lint warnings as shown inline with this PR diff)

You can run `make dev` to test and see your results.

fiunchinho commented:
> LGTM, but the build still fails (wrong `impl-title`, Markdown lint warnings as shown inline with this PR diff)
> 
> You can run `make dev` to test and see your results.

Fixed, and
- added note about renaming node pools
- added node pool deletion
